bootstrap/redactor.go

```diff
@@ -0,0 +1,197 @@
+			}
+			if confirmedTo > doneTo {
+				redactor.outbuf = append(redactor.outbuf, input[doneTo:confirmedTo]...)
+				doneTo = confirmedTo
+			}
+
+			continue
+		}
+
+		// We'll check for matching search strings here, but we'll still need
+		// to move the cursor forward
+		// Since Go slice syntax is not inclusive of the end index, moving it
+		// forward now reduces the need to use `cursor-1` everywhere
+		cursor++
+		for _, needle := range redactor.table[ch].needles {
+			// Since we're working backwards from what may be the end of a
+			// string, it's possible that the start would be out of bounds
+			startSubstr := cursor - len(needle)
+			var candidate []byte
+
+			if startSubstr >= 0 {
+				// If the candidate string falls entirely within input, then just slice into input
+				candidate = input[startSubstr:cursor]
+			} else if -startSubstr < len(redactor.outbuf) {
+				// If the candidate crosses the Write boundary, we need to
+				// concatenate the two sections to compare against
+				candidate = make([]byte, 0, len(needle))
+				candidate = append(candidate, redactor.outbuf[-startSubstr:]...)
+				candidate = append(candidate, input[:cursor]...)
+			} else {
+				// Final case is that the start index is out of bounds, and
+				// it's impossible for it to match. Just move on to the next
+				// search substring
+				continue
+			}
+
+			if bytes.Equal(needle, candidate) {
+				if startSubstr < 0 {
+					// If we accepted a negative startSubstr, the output buffer
+					// needs to be truncated to remove the partial match
+					redactor.outbuf = redactor.outbuf[:len(redactor.outbuf)+startSubstr]
+				} else if startSubstr > doneTo {
+					// First, copy over anything behind the matched substring unmodified
+					redactor.outbuf = append(redactor.outbuf, input[doneTo:startSubstr]...)
+				}
+				// Then, write a fixed string into the output, and move doneTo past the redaction
+				redactor.outbuf = append(redactor.outbuf, redactor.replacement...)
+				doneTo = cursor
+
+				// The next end-of-string will be at least this far away so
+				// it's safe to skip forward a bit
+				cursor += redactor.minlen-1
+				break
+			}
+		}
+	}
+
+	// Push the output buffer down
+	_, err := redactor.output.Write(redactor.outbuf)
+
+	// There will probably be a segment at the end of the input which may be a
+	// partial match crossing the Write boundary. This is retained in the
+	// output buffer to compare against on the next call
+	// Flush() needs to be called after the final Write(), or this bit won't
+	// get written
+	redactor.outbuf = append(redactor.outbuf[:0], input[doneTo:]...)
+
+	// We can offset the next Write processing by how far cursor is ahead of
+	// the end of this input segment
+	redactor.offset = cursor - len(input)
+
+	return len(input), err
+}
+
+// Flush should be called after the final Write. This will Write() anything
+// retained in case of a partial match and reset the output buffer.
+func (redactor Redactor) Sync() error {
```

Comment:
I tend to think `Flush` was the better method name, did the semantics change to make `Sync` more appropriate?

Reply:
Because naming things is hard. I was thinking of the `sync`/`fsync` system calls in linux for triggering an immediate write of the file system buffers. However, I just noticed there's also an `fflush` libc function for doing almost the same thing (write the libc buffer to the operating system).

I'm happy to change it back.